src/test/librbd/image/test_mock_RemoveRequest.cc

```diff
@@ -0,0 +1,315 @@
+// -*- mode:C; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "test/librbd/test_mock_fixture.h"
+#include "test/librbd/test_support.h"
+#include "test/librbd/mock/MockImageCtx.h"
+#include "test/librbd/mock/MockContextWQ.h"
+#include "test/librados_test_stub/MockTestMemIoCtxImpl.h"
+#include "test/librados_test_stub/MockTestMemRadosClient.h"
+#include "librbd/ImageState.h"
+#include "librbd/internal.h"
+#include "librbd/Operations.h"
+#include "librbd/operation/TrimRequest.h"
+#include "librbd/operation/DisableFeaturesRequest.h"
+#include "librbd/image/RemoveRequest.h"
+#include "librbd/image/RefreshParentRequest.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include <arpa/inet.h>
+#include <list>
+#include <boost/scope_exit.hpp>
+
+namespace librbd {
+namespace operation {
+
+template <>
+class TrimRequest<MockImageCtx> {
+public:
+  static TrimRequest *s_instance;
+  static TrimRequest *create(MockImageCtx &image_ctx, Context *on_finish,
+                             uint64_t original_size, uint64_t new_size,
+                             ProgressContext &prog_ctx) {
+    assert(s_instance != nullptr);
+    s_instance->on_finish = on_finish;
+    return s_instance;
+  }
+
+  Context *on_finish = nullptr;
+
+  TrimRequest() {
+    s_instance = this;
+  }
+
+  MOCK_METHOD0(send, void());
+};
+
+template <>
+class DisableFeaturesRequest<MockImageCtx> {
+public:
+  static DisableFeaturesRequest *s_instance;
+  static DisableFeaturesRequest *create(MockImageCtx &image_ctx,
+                                                  Context *on_finish,
+                                                  uint64_t journal_op_tid,
+                                                  uint64_t features,
+                                                  bool force) {
+    assert(s_instance != nullptr);
+    s_instance->on_finish = on_finish;
+    return s_instance;
+  }
+
+  Context *on_finish = nullptr;
+
+  DisableFeaturesRequest() {
+    s_instance = this;
+  }
+
+  MOCK_METHOD0(send, void());
+};
+
+DisableFeaturesRequest<MockImageCtx> *DisableFeaturesRequest<MockImageCtx>::s_instance = nullptr;
+} // namespace operation
+} // namespace librbd
+
+// template definitions
+#include "librbd/image/RemoveRequest.cc"
+template class librbd::image::RemoveRequest<librbd::MockImageCtx, librbd::MockContextWQ>;
+#include "librbd/operation/TrimRequest.cc"
```

Comment:
Shouldn't need to include the definition to the TrimRequest template -- use template specialization to eliminate it.